### ISSUE TITLE
[Routing] Do not let `RouteCollection` override `Route` conditions

### ISSUE THREAD

**Issue by mbuliard** [symfony/symfony#53171]
### Description

Hi,

**TLDR**: General Condition defined in YAML/XML collections overrides specific condition defined in attribute.

I have a trouble with route conditions on a project I'm working on.
Routes are loading via a mix of YAML and Attributes :
```yaml
api:
    resource: ../../src/Controller/
    type: annotation
    condition: "%general_condition%"
```
```php
#[Route(path: "my/route", condition: "specific condition")]
final class MyController {
   ...
}
```
My trouble is that the specific condition, defined by attribute, is overrided by the general condition, defined via YAML.
Looking into the component, I found out that when the YamlFileLoader sets the condition on the RouteCollection, the RouteCollection overrides any previous condition.
I find it disturbing that a general condition overrides a specific condition.
```php
public function setCondition(?string $condition)
{
    foreach ($this->routes as $route) {
        $route->setCondition($condition);
    }
}
```

**Propositions**
1) Make specific overrides general, but I fear it is a BC since previously enforced general condition would not be any more.
```php
public function setCondition(?string $condition)
{
    foreach ($this->routes as $route) {
        if ('' === $route->getCondition()) {
            $route->setCondition($condition);
        }
    }
}
```

2) Combine conditions. I don't think of it as a BC since one cannot expect its attribute condition to **not** work.
```php
public function setCondition(?string $condition)
{
    foreach ($this->routes as $route) {
        if ('' === $route->getCondition()) {
            $route->setCondition($condition);
        } else {
           $route->setCondition('(' . $route->getCondition . ') and (' . $condition . ')');
        }
    }
}
```

**Comment by marac19901990** [comment]
From the code above for me it makes sense it doesn't use the attribute from the controller but the yaml general config because you set the type to annotation in config. 
So it will be looking for annotated routes, and it won't look at attributes since not specified.
If I didn't see something and am not right I'll be happy to try everything out and try to help you.

**Comment by marac19901990** [comment]
I was wrong, I managed to reproduce your issue, the type doesn't change anything. Will try to help you figure something out.

**Comment by marac19901990** [comment]
I've spent some time looking at potential solutions. And I must say that I wasn't able to find an elegant one worth making a pull request. The code is also a bit all over the place, there are similar methods that are used arbitrarily seemed to me. 
Maybe a stupid remark and someone has a great solution. 
**What i would do here is remove the option to set the condition globally from the config file.**  
The reasoning behind the idea being that it would further complicate the code that already is complicated and from my point of view it is not a big deal to do it per class. If anyone has a better solution I will be happy that I was wrong, and would just embrace the learning.

**Comment by mbuliard** [comment]
Hello,
I let the subject down during Christmas season, but here is a PR fixing the issue : https://github.com/symfony/symfony/pull/53511

@marac19901990 Thanks for your time. I think the merging of the global and specific condition is an acceptable solution. Letting the specific override the global would be also acceptable in my opinion, but may be a breaking change.

**Comment by stof** [comment]
This is not only for the condition. Currently, requirements and defaults also work the same: a valid set on the import wins over the config defined inside the imported resource. Reversing the behavior is a BC break

**Comment by mbuliard** [comment]
> This is not only for the condition. Currently, requirements and defaults also work the same: a valid set on the import wins over the config defined inside the imported resource. Reversing the behavior is a BC break

Thanks for you reply.
The current behavior of 'global overrides specific' seems weird to me.
Do you think it should be kept or a migration path, maybe via an option, should be done ?
